STN-661: Carousel Height Refinement

heatherdesigns opened the pull request:
# [STN-661](https://sparkbox.atlassian.net/browse/STN-661) READY FOR REVIEW

## Summary
Part One:

The work in this card refines the height of text boxes in the Carousel. JS is used to find the height of each text area and identify the tallest. The tallest text area is used to update all containers to match that height.

Part Two:

For site owners that would prefer the existing behavior a utility class called `hb-carousel-reset-height` has been created to prevent the JS from running for that carousel (or any carousels in the same container).

Notes: 
* It was determined not to include this work within the experimental toggle. Reference [slack thread](https://sparkbox.slack.com/archives/CM9J5THPV/p1613675494110900)

## Need Review By (Date)
2/26/2021

## Urgency
medium

## Steps to Test
1. In your CLI, run `npm run test` and confirm all tests pass. Then run `npm run build` to build the CSS and JS.
2. To confirm the height of the text boxes in carousels are all set to the same height as the carousel with the tallest text box height, go to http://sparkbox-sandbox.suhumsci.loc/carousel-hero-image. You if do not have this page locally, you can sync with staging or create a new flexible page and add a carousel with multiple slides in the hero area and again within the main content section. If you sync with staging, note that you may need to upload new images for each carousel.
3. - [x] Confirm the text boxes on multiple slides within a single carousel are the same height on both Colorful and Traditional themes.
4. - [x] Resize the width of the browser window. Confirm the text boxes resize according the tallest box with the most content. Note the slick dots and arrows have always had some alignment funkiness when the page is resized. This is pre-existing.
5. - [x] Test new utility class of `hb-carousel-reset-height`. Add `hb-carousel-reset-height` to a carousel component via Layout Builder. Confirm that the height of the text boxes for each slide within a carousel defaults to fit the content rather than the tallest text box.
6. - [x] Confirm that the JavaScript only runs on pages with carousels.
7.  - [x] Confirm there are not JavaScript errors from the work done in this PR. (There are some errors re: fontawesome and a pre-existing  Uncaught TypeError.)
8. - [x] Confirm the work in this PR does not impact other hero elements.
9. - [ ] Confirm the new utility class of `hb-carousel-reset-height` has been added to the docs.

Commit:
feat (STN-661): Carousel Height Refinement

Commit:
feat: parse through JS to find the heights of the boxes which contain content inside of carousels

Commit:
feat: add JS timeout functions to prevent the window event listener from firing the restrictHeight function until the window is no longer being resized

Commit:
feat: add hb-carousel-reset-height utility class to docs

Commit:
feat: prevent window resize event listener from listening on pages that do not have carousel slides

heatherdesigns commented on docroot/themes/humsci/humsci_basic/docs/utility-classes.md:
```diff
@@ -43,7 +43,7 @@ Field Utility Classes are used to apply styles to a field in a view (Drupal Home
 | hb-heading-5         | Applies heading 5 styles | ✅|
 | hb-heading-6         | Applies heading 6 styles | ✅|
-| hb-heading-link-color | Applies link color styles to a linked heading. Do not use with dark color inversion toggle or hb-dark-pattern. | ✅|
+| hb-heading-link-color | Applies link color styles to a linked heading. Do not use with hb-dark-pattern. | ✅|
```
I took the opportunity to update the description for hb-heading-link-color.

kaseybon replied:
👏

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+  restrictHeight();
+
+  // Watch for when the browser window resizes, then run the restrictHeight
+  // function to reset the height of the text boxes
```
I added a lot of comments thinking of how I could make it easier for Ryan to review. Please let me know if there are any that I should remove for being too verbose.

kaseybon replied:
This is great! It makes the code more accessible for anyone who is reviewing.

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+
+    // Clear any inline styles that may have been set previously
+    // This is necessary to determine the default height of text boxes
+    for (let n = 0; n < textBoxes.length; n++) {
```
Can we combine this for loop with the one on line 31?

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+      boxHeight = parseInt(boxHeight, 10);
+
+      // Create an array containing all the heights of textBoxes
+      boxHeightArray.push(boxHeight);
```
Because of all the nested for loops this is running 3 times for each slide. This is going to have a negative effect on performance.

heatherdesigns replied:
I'm not sure I understand. I did a console log of boxHeightArray and only see it running when a new height value is added to the array.

![push](https://user-images.githubusercontent.com/12678977/109857065-3e424a80-7c28-11eb-9c1a-6e57030d6823.png)

kaseybon replied:
Ah, that makes sense. I was forgetting how loops work 🤦🏻‍♀️

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+function restrictHeight() {
+  for (let i = 0; i < slides.length; i++) {
+    let textBoxes;
+    let boxHeightArray = [0]; // array must have a default entry of 0 for the banner components
```
We should pull `box-HeightArray` outside of the for loop. We don't need to redeclare it each time the code runs.

heatherdesigns replied:
I moved the initial declaration outside of the for loop, but it needs to be set at [0] inside the loop to set a new baseline for each time the code runs for each carousel on a page.

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+// Set the height of all text boxes within a Carousel to that
+// of the tallest text box
+function restrictHeight() {
+  for (let i = 0; i < slides.length; i++) {
```
The [forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) loop is a bit easier to read and we won't need to declare all of these counter variables.

heatherdesigns replied:
For some reason I thought forEach was not supported in IE and chose for loops instead. forEach are definitely easier to implement. :smile:

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+      let boxHeight = textBoxes[j].offsetHeight;
+
+      // Parse boxHeight to be a number that can be used to set the min-height value
+      boxHeight = parseInt(boxHeight, 10);
```
I don't think you need the second argument in `parseInt`.

kaseybon commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,52 @@
+    let maxBoxHeight = Math.max(...boxHeightArray);
+
+    // Give all textBoxes the same height on medium and larger sized screens
+    if (window.innerWidth > 768) {
```
Instead of doing this check here we should do it before we even run the function. That way we don't go through all of the work of finding the heights if the screen is too small. 

Can we also set 768 as a variable to describe what breakpoints we're targeting?

Commit:
feat: begin implementing review feedback

- add variable for medium screen breakpoint to descibe the number we are targeting
- remove unnecessary number from parseInt function
- refactor when the restrictHeight function runs to prevent it from running at smaller screen sizes

Commit:
feat: refactor 2 loops into 1 loop to improve performance

Commit:
feat: remove boxHeightArray variable declaration from within the for loop to prevent it from being redeclared each time the code runs

Commit:
fix: update boxHeightArray variable declaration to allow a baseline value to be set for each individual carousel on a page

Commit:
feat: refactor for loops into forEach loops

kaseybon reviewed (approved):
This looks great! Awesome work!

![giphy (6)](https://user-images.githubusercontent.com/2486846/109983148-8f0d7e00-7cd0-11eb-86b7-6f0ffaee4fac.gif)

Commit:
fix: hb-carousel-reset-height utility class to function correctly

Commit:
feat: add compiled CSS & JS

heatherdesigns commented on docroot/themes/humsci/humsci_basic/src/js/shared/carousel-slides/carousel-slides-height.js:
```diff
@@ -0,0 +1,57 @@
+  window.addEventListener('resize', function() {
+    if (window.innerWidth > mediumScreenBreakpoint) {
+      clearTimeout(timeOutFunctionId);
+      timeOutFunctionId = setTimeout(restrictHeight, 250);
```
I think we may need to remove the time out functions to resolve the issue with the slick dots/arrows that Tori mentioned in her feedback in [STN-661](https://sparkbox.atlassian.net/browse/STN-661).
